### PR TITLE
all: Convert some standard errors into well known errors

### LINKED ISSUE #2448: Convert common/well-known errors to defined errors
#### Summary

We should convert some common and well-known errors to defined errors.

#### Why do we need this?

To give more insight into what is going on when an error is returned.

#### What is already there? What do you see now?

If during OAuth token exchange a DNS error, timeout or other network error occurs, an internal server error is returned without details.

#### What is missing? What do you want to see?

Some details about the network error. Maybe DNS couldn't be resolved? Maybe the TCP connection timed out? Maybe the TLS expired? ...

#### How do you propose to implement this?

We could make `errors.From` recognize some common errors and return rich errors with attributes (built from definitions).

#### Can you do this yourself and submit a Pull Request?

Yes, but if anyone else would like to do this, just comment and re-assign.

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #2448

#### Changes

- Well defined errors for context canceled and deadline exceeded
- `errors.From` can parse most common net.Errors

#### Testing

unit tests

##### Regressions

...

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [X] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/errors/errors_test.go [hunk 1]
```diff
@@ -74,3 +77,152 @@ func TestFields(t *testing.T) {
 		"error_cause_cause": "back",
 	})
 }
+
+func TestContextCancelled(t *testing.T) {
```
Comment (johanstokking) [inline pkg/errors/errors_test.go:81]: Canceled

File: pkg/errors/conversion.go [hunk 2]
```diff
@@ -65,6 +86,38 @@ func From(err error) (out *Error, ok bool) {
 			e = e.WithCause(cause)
 		}
 		return &e, true
+	case *net.DNSError:
+		e := build(errNetDNS, 0).WithAttributes(
+			append([]interface{}{"not_found", err.IsNotFound}, netErrorDetails(err)...)...,
```
Comment (johanstokking) [inline pkg/errors/conversion.go:91]: You should be able to chain the call with `WithAttributes()`

File: pkg/errors/conversion.go [hunk 3]
```diff
@@ -15,10 +15,24 @@
...
 
+var (
+	errContextDeadlineExceeded = DefineDeadlineExceeded("context_deadline_exceeded", "context deadline exceeded")
+	errContextCanceled         = DefineCanceled("context_canceled", "context canceled")
+
+	errNetInvalidAddr    = DefineInvalidArgument("net_invalid_addr", "invalid address error: `{message}`", "message", "temporary", "timeout")
+	errNetAddr           = DefineUnavailable("net_addr", "address error: `{message}`", "message", "temporary", "timeout")
+	errNetDNS            = DefineUnavailable("net_dns", "DNS error: `{message}`", "message", "temporary", "timeout", "not_found")
+	errNetUnknownNetwork = DefineNotFound("net_unknown_network", "unknown network: `{message}`", "message", "temporary", "timeout")
+	errNetOperation      = DefineUnavailable("net_operation", "network operation error: `{message}`", "message", "op", "net", "source", "address", "timeout", "temporary")
```
Comment (johanstokking) [inline pkg/errors/conversion.go:33]: - We need the `error:` text in the error message
- You don't need to use backticks for the error message
- We can also consider just using the error message as-is, and use the error name to know what kind of error it is

**Comment by neoaggelos** [comment]
> Why is this targeting `develop`?

This is possibly changing errors shown in the console or returned by the API. Is this not a breaking change?

**Comment by johanstokking** [comment]
> This is possibly changing errors shown in the console or returned by the API. Is this not a breaking change?

No, that is not a breaking change. See the compatibility commitment; it does not explicitly state errors

**Comment by htdvisser** [comment]
@neoaggelos Travis is satisfied. You can merge this.

**Pull request merged**